### PR TITLE
Third-party license attribution

### LINKED ISSUE #19: Add view of third-party licenses
Need a view you can open to show all third-party licenses from.

This is a requirement for us to be able to ship the code that we're using.

This also includes doing a small bit of investigation of what licenses our dependencies actually use. As much as I like copy-left, the rest of the company disagrees and so if there are any copy-left licenses then we must remove those dependencies.

Only the regular prod dependencies are needed. The `devDependencies` may be ignored, as we are not shipping that code with the product.

As this list easily gets outdated, suggest to make it generated. Suggest using something like [davglass/license-checker](https://github.com/davglass/license-checker) outputted as JSON to the assets directory and then when rendering it we link to the SPDX pages for each license (ex: https://spdx.org/licenses/MIT). It's not foolproof, but it's better than nothing

The licenses page could be linked to in the version panel found below the "WHARF" header in the sidebar, so for example:
Screenshot_2021-05-25_10-36-49

### PR DESCRIPTION
**Pull request by jilleJr**
- \[x] I've added a new note in the `CHANGELOG.md` file, according to docs:
  https://iver-wharf.github.io/#/development/changelogs/writing-changelogs

## Summary

- Added node.js script `collect-licenses.js` to obtain all licenses using [`wharf-collect-licenses`](https://github.com/iver-wharf/wharf-collect-licenses-js)
- Changed so `build-prod` also runs `collect-licenses.js` node script
- Added LicensesService to obtain generated `licenses.json` asset
- Added `/third-party-licenses` page
- Updated tsconfig with strictTemplates for Angular 9
- Overwrote PrimeNG's Accordion styling

## Motivation

Closes #19

## Preview

Screenshot from 2021-09-17 11-22-20

Screenshot from 2021-11-05 10-27-58

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -5,8 +5,9 @@
     "ng": "ng",
     "start": "ng serve",
     "build": "ng build",
-    "build-prod": "ng build -c production",
+    "build-prod": "\"$npm_execpath\" run collect-licenses && ng build -c production",
```
Comment (fredx30) [inline package.json:8]: Would this work for windows environments?

**Review by fredx30 — COMMENTED** [review]
This adds alot of code that i dont really think belongs to the core wharf codebase. Could this perhaps be moved out to an angular library?

**Comment by jilleJr** [comment]
> @fredx30 This adds alot of code that i dont really think belongs to the core wharf codebase. Could this perhaps be moved out to an angular library?

And by "this" you're referring to the `collect-licenses` script? Sure we could extract it into it's own repo and then into its own NPM package, but that feels like overkill. It would be its own command-line utility.
It's not an angular library though, so maybe you are referring to something else, then what?

File: package.json (hunk 1 above)
Comment (jilleJr) [reply package.json:8]: On powershell, probably, but TIL that if you're using batch then it wants the `%npm_execpath%` syntax... Sigh

I'll remove this and just call `npm run collect-licenses` manually in the docker image instead.

**Comment by fredx30** [comment]
> > @fredx30 This adds alot of code that i dont really think belongs to the core wharf codebase. Could this perhaps be moved out to an angular library?
> 
> And by "this" you're referring to the `collect-licenses` script? Sure we could extract it into it's own repo and then into its own NPM package, but that feels like overkill. It would be its own command-line utility.
> It's not an angular library though, so maybe you are referring to something else, then what?

Im thinking that it could be a standalone component library. It could be added to a template such as `<licenses-list></licenses-list>`. 
If it could be made into a standalone component then it wouldnt have to add build steps, im refering to the collect licenses step here.  It wouldnt have to add thirdparty dependencies to the core package, it would do so indrectly which im still not a big fan of but atleast that way its modular in the sense that if the requirement goes the dependency does as well. It wouldnt add the task of solving administrative issues such as licensing to the core codebase.

**Comment by jilleJr** [comment]
Had a verbal meeting with @fredx30, and I researched a little further. Here are my conclusions:

### Why embed, instead of linking to licenses

We need the list of dependencies and their licenses, so `collect-licenses` is needed nonetheless. Question is if we need the license texts, or can we just link to the licenses?

Legally, I was not able to find any concrete answers to this. Best I could find was this: https://www.jolts.world/index.php/jolts/article/view/47/102 which analyses the "license attribution" requirements of many common licenses, including MIT and Apache-2.0 that we depend on here in wharf-web. The conclusion from that article was that none actually require the license text to be embedded, and we could *probably* settle with only linking to the license texts.

Technically, obtaining the packages through `npm run collect-licenses` takes:

- ~5.5 - 6 sec to only get the package names and their SPDX license names
- ~6 - 7 sec to get the packages names, their SPDX license names, and their license texts by obtain some from GitHub.com where they are missing from their NPM packages.

What takes time is the TypeScript compilation and my slow harddisk drive. Above results are without any cached build files (removed `deploy/collect-licenses/dist` between each benchmark test). Worth noting again that this only affects the production builds, so this adds 7 seconds to the already 3-4 minutes long build time of the Angular production build, while this is not required for local development so it adds 0 seconds to the development iteration time.

So in the spirit of "better safe than sorry", I conclude that we should keep the license texts in the built product.

### Extracting to a separate package

Pros:

- Can precompile the typescript, reducing time taken to run `npm run collect-licenses`
- Less bloat in `package.json`
- Reusable, could be useful in other non-wharf projects as well

Cons:

- Takes time to dev
- More packages => more to maintain

The pros outweighs the cons, so I'll extract this into its own package.

### Does licensing considerations belong in the repo

No better place to put it. Will extract into its own package, as mentioned above, but that's just for de-bloating the repo.

I've no better answer than that. If you have suggestions, I'm all ears.

**Comment by jilleJr** [comment]
I've extracted the code into the new repo https://github.com/iver-wharf/wharf-collect-licenses-js and as it's pre-compiled TypeScript code the execution time of `npm run collect-licenses` has dropped down to around ~ 3 - 4 seconds.

It's now ready for review again.

**Review by Alexamakans — APPROVED** [review]

**Review by fredx30 — APPROVED** [review]
Good work!

**Pull request merged**